### PR TITLE
Add deeplyEqual() method for use in replay

### PR DESCRIPTION
**Pull request by tvolkert**
Part of #11

### CODE REVIEW

File: lib/src/backends/record_replay/common.dart [hunk 1]
```diff
@@ -83,3 +87,40 @@ class TypeMatcher<T> {
...
+///
+/// Two lists are deeply equal if they have the same type, the same length, and
+/// every element in list A is pairwise deeply equal with the corresponding
+/// element in list B.
+///
+/// Two maps are deeply equal if they have the same type, the same length, the
+/// same set of keys, and the value for every key in map A is deeply equal to
+/// the corresponding value in map B.
+///
+/// All other types of objects are deeply equal if they are logically equal.
```
Comment (cbracken) [inline lib/src/backends/record_replay/common.dart]: I would be pedantic and make clear that they have to be of exactly the same runtime type, not just is-a.

File: test/recording_test.dart [hunk 2]
```diff
@@ -245,6 +245,67 @@ void main() {
...
+        expect(deeplyEqual(<String, int>{}, null), isFalse);
+      });
+
+      test('listOfMaps', () {
+        expect(deeplyEqual(newMap(), newMap()), isTrue);
+        expect(deeplyEqual(newMap(), newMap(stringValue: 'bar')), isFalse);
+        expect(deeplyEqual(newMap(), newMap(boolValue: false)), isFalse);
+        expect(deeplyEqual(newMap(), newMap(lastListValue: 'd')), isFalse);
+        expect(deeplyEqual(newMap(), newMap(lastMapValue: 3)), isFalse);
+      });
```
Comment (cbracken) [inline test/recording_test.dart]: I would add tests that check that lists, maps and some object type that are is-a conformant and == conformant fail due to the runtimeType check.

**Comment by cbracken** [comment]
lgtm

File: test/recording_test.dart (hunk 2 above)
Comment (tvolkert) [inline test/recording_test.dart]: I already have that for list (`expect(deeplyEqual(<int>[], <dynamic>[]), isFalse);`) - I'll add it for map

**Pull request merged**